drivers/gpio/gpio_stmpe1600.c

```diff
@@ -0,0 +1,324 @@
+#define REG_IEGPIOR_LSB 0x08    /* GPIO interrupt enable register */
+#define REG_IEGPIOR_MSB 0x09
+#define REG_ISGPIOR_LSB 0x0A    /* GPIO interrupt status register */
+#define REG_ISGPIOR_MSB 0x0B
+#define REG_GPMR_LSB    0x10    /* GPIO monitor pin state register */
+#define REG_GPMR_MSB    0x11
+#define REG_GPSR_LSB    0x12    /* GPIO set pin state register */
+#define REG_GPSR_MSB    0x13
+#define REG_GPDR_LSB    0x14    /* GPIO set pin direction register */
+#define REG_GPDR_MSB    0x15
+#define REG_GPPIR_LSB   0x16    /* GPIO polarity inversion register */
+#define REG_GPPIR_MSB   0x17
+
+
+/** Configuration data */
+struct stmpe1600_config {
+	/* gpio_driver_config needs to be first */
+	struct gpio_driver_config common;
+
+	/** Master I2C device */
+	const struct device *i2c_bus;
+
+	/** The slave address of the chip */
+	uint16_t i2c_slave_addr;
+};
+
+/** Runtime driver data */
+struct stmpe1600_drvdata {
+	/* gpio_driver_data needs to be first */
+	struct gpio_driver_data common;
+
+	/** Driver lock */
+	struct k_sem lock;
+
+	/* Registers cache */
+	uint16_t GPSR;
+	uint16_t GPDR;
+};
+
+static int write_reg16(const struct stmpe1600_config * const config, uint8_t reg, uint16_t value)
+{
+	int ret;
+	uint16_t transfer_data = sys_cpu_to_le16(value);
+
+	LOG_DBG("STMPE1600[0x%02X]: write REG[0x%02X..0x%02X] = %04x",
+		config->i2c_slave_addr, reg, reg + 1, value);
+
+	ret = i2c_burst_write(config->i2c_bus, config->i2c_slave_addr, reg,
+			      (uint8_t *)&transfer_data, sizeof(transfer_data));
+
+	if (ret != 0) {
+		LOG_ERR("STMPE1600[0x%02X]: write error REG[0x%02X..0x%02X]: %d",
+			config->i2c_slave_addr, reg, reg + 1, ret);
+	}
+	return ret;
+}
+
+static int read_reg16(const struct stmpe1600_config * const config, uint8_t reg, uint16_t *value)
+{
+	int ret;
+	uint16_t transfer_data;
+
+	LOG_DBG("STMPE1600[0x%02X]: read REG[0x%02X..0x%02X]",
+		  config->i2c_slave_addr, reg, reg + 1);
+
+	ret = i2c_burst_read(config->i2c_bus, config->i2c_slave_addr, reg,
+			     (uint8_t *) &transfer_data, sizeof(transfer_data));
+
+	if (ret != 0) {
+		LOG_ERR("STMPE1600[0x%02X]: read error REG[0x%02X..0x%02X]: %d",
+			config->i2c_slave_addr, reg, reg + 1, ret);
+	} else {
+		*value = sys_le16_to_cpu(transfer_data);
+		LOG_DBG("STMPE1600[0x%02X]: read REG[0x%02X..0x%02X] => %04x",
+			config->i2c_slave_addr, reg, reg + 1, *value);
+	}
+	return ret;
+}
+
+static int set_pin_dir(const struct device *dev, gpio_pin_t pin, gpio_flags_t flags)
+{
+	int ret;
+	bool set_state = false;
+	struct stmpe1600_drvdata *const drvdata = (struct stmpe1600_drvdata *const) dev->data;
+	uint16_t GPDR = drvdata->GPDR;
+	uint16_t GPSR = drvdata->GPSR;
+
+	if ((flags & GPIO_INPUT) != 0U) {
+		GPDR &= ~BIT(pin);
+	}
+	if ((flags & GPIO_OUTPUT) != 0U) {
+		GPDR |= BIT(pin);
+		if ((flags & GPIO_OUTPUT_INIT_HIGH) != 0U) {
+			GPSR |= BIT(pin);
+			set_state = true;
+		} else if ((flags & GPIO_OUTPUT_INIT_LOW) != 0U) {
+			GPSR &= ~BIT(pin);
+			set_state = true;
+		}
+	}
```

Comment:
It would be cleaner if we ignored `GPIO_INPUT` flag and added `else` branch here. I.e. let's remove 
```
	if ((flags & GPIO_INPUT) != 0U) {
		GPDR &= ~BIT(pin);
	}
```
and instead add here
```
	} else {
		GPDR &= ~BIT(pin);
	}
```
Since it's in fact `GPIO_OUTPUT` flag that exclusively controls the setting of the GPDR register.